### PR TITLE
[wasm] Implement MONO_MEMORY_BARRIER in jiterpreter

### LINKED ISSUE #100273: [browser][MT] enable jiterp
I [disabled jiterp](https://github.com/dotnet/runtime/pull/100156) to simplify troubleshooting of [deadlocks](https://github.com/dotnet/runtime/issues/99888)

We should enable it again after we figure out the root cause of the deadlocks.

### PR DESCRIPTION
**Pull request by kg**
This opcode is fairly high in the abort statistics for System.Runtime.Tests, so it's worth the trouble to wire it up

Fixes https://github.com/dotnet/runtime/issues/100273

### CODE REVIEW

**Comment by jkotas** [comment]
Should this be just a no-op in single threaded wasm?

**Comment by pavelsavara** [comment]
Right now jiterp is disabled in MT, so we need to enable it in order to test this PR.
https://github.com/dotnet/runtime/issues/100273

https://github.com/dotnet/runtime/blob/6e68f9233bb408e2862251858aabe87dd68532fd/src/mono/mono/utils/options-def.h#L92

**Comment by kg** [comment]
I'll do a few runs with traces enabled for MT tomorrow.

File: src/mono/browser/runtime/jiterpreter-trace-generator.ts [hunk 1]
```diff
@@ -3953,6 +3954,12 @@ function emit_simd_4 (builder: WasmBuilder, ip: MintOpcodePtr, index: SimdIntrin
 function emit_atomics (
     builder: WasmBuilder, ip: MintOpcodePtr, opcode: number
 ) {
+    if (opcode === MintOpcode.MINT_MONO_MEMORY_BARRIER) {
+        if (WasmEnableThreads)
+            builder.callImport("membar");
+        return true;
+    }
+
```
Comment (lambdageek) [inline src/mono/browser/runtime/jiterpreter-trace-generator.ts:3962]: Why? `mono_memory_barrier` is just `__sync_synchronize` (it's an ABI memory barrier, not a GC write barrier), so you could just emit `atomic.fence` directly, no?

https://godbolt.org/z/jjWsxzj9h

**Review by lambdageek — APPROVED** [review]

**Review by pavelsavara — APPROVED** [review]

File: src/mono/browser/runtime/jiterpreter-trace-generator.ts [hunk 2]
```diff
@@ -3977,8 +3977,12 @@ function emit_atomics (
     builder: WasmBuilder, ip: MintOpcodePtr, opcode: number
 ) {
     if (opcode === MintOpcode.MINT_MONO_MEMORY_BARRIER) {
-        if (WasmEnableThreads)
+        if (WasmEnableThreads) {
             builder.appendAtomic(WasmAtomicOpcode.atomic_fence);
+            // The text format and other parts of the spec say atomic.fence has no operands,
+            //  but the binary encoding requires a dummy zero byte for some reason
```
Comment (lambdageek) [inline src/mono/browser/runtime/jiterpreter-trace-generator.ts:3983]: It's for the memory ordering. See this part of the shared everythign threads proposal. 0 means sequential consistency, 1 will mean acquire-release https://github.com/WebAssembly/shared-everything-threads/blob/main/proposals/shared-everything-threads/Overview.md#memory-orderings

Comment (lambdageek) [reply src/mono/browser/runtime/jiterpreter-trace-generator.ts]: > And sync_synchronize is sequential consistency, right?

It's a "full barrier" which presumably means seq cst. https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html

**Pull request merged**